src/v/iceberg/merge_append_action.cc

```diff
@@ -351,6 +351,160 @@ merge_append_action::upload_as_manifest(
     co_return co_await io_.upload_manifest(path, m);
 }
 
+namespace {
+
+struct promoting_visitor {
```

Comment:
I guess we don't actually promote types but just check validity in the schema evolution work @oleiman has been working on.

Can we at least consolidate this to `compatibility.cc` and add some tests?

Reply:
We do promote int -> long and float -> double values. I think these are the only cases relevant for partition keys in v2 iceberg (there are also decimals, but the underlying variant is the same in our impl).

Reply:
iiuc, this is for rejiggering the partition tuples on data files that were written before the schema changed. is that about right?

Reply:
> iiuc, this is for rejiggering the partition tuples on data files that were written before the schema changed. is that about right?

yes, exactly